network/src/noise_wrapper/handshake.rs

```diff
@@ -60,8 +60,59 @@ impl AntiReplayTimestamps {
 /// but as we use it to store a duration since UNIX_EPOCH we will never use more than 8 bytes.
 const PAYLOAD_SIZE: usize = 8;
 
-// Noise Wrapper
-// -------------
+/// Noise handshake authentication mode.
+pub enum HandshakeAuthMode {
+    /// In `Mutual` mode, both sides will authenticate each other with their
+    /// `trusted_peers` set. We also include replay attack mitigation in this mode.
+    ///
+    /// For example, in the Libra validator network, validator peers will only
+    /// allow connections from other validator peers. They will use this mode to
+    /// check that inbound connections authenticate to a network public key
+    /// actually contained in the current validator set.
+    Mutual {
+        // Only use anti replay protection in mutual-auth scenarios. In theory,
+        // this is applicable everywhere; however, we would need to spend some
+        // time making this more sophisticated so it garbage collects old
+        // timestamps and doesn't use unbounded space. These are not problems in
+        // mutual-auth scenarios because we have a bounded set of trusted peers
+        // that rarely changes.
+        anti_replay_timestamps: RwLock<AntiReplayTimestamps>,
+        trusted_peers: Arc<RwLock<HashMap<PeerId, NetworkPeerInfo>>>,
+    },
+    /// In `ServerOnly` mode, the dialer authenticates the server. However, the
+    /// server does not care who connects to them and will allow inbound connections
+    /// from any peer.
+    ServerOnly,
+}
+
+impl HandshakeAuthMode {
+    pub fn mutual(trusted_peers: Arc<RwLock<HashMap<PeerId, NetworkPeerInfo>>>) -> Self {
+        HandshakeAuthMode::Mutual {
+            anti_replay_timestamps: RwLock::new(AntiReplayTimestamps::default()),
+            trusted_peers,
+        }
+    }
+
+    fn anti_replay_timestamps(&self) -> Option<&RwLock<AntiReplayTimestamps>> {
+        match &self {
+            HandshakeAuthMode::Mutual {
+                anti_replay_timestamps,
+                ..
+            } => Some(&anti_replay_timestamps),
+            HandshakeAuthMode::ServerOnly => None,
+        }
+    }
+
+    fn trusted_peers(&self) -> Option<&RwLock<HashMap<PeerId, NetworkPeerInfo>>> {
+        match &self {
+            HandshakeAuthMode::Mutual { trusted_peers, .. } => Some(&trusted_peers),
+            HandshakeAuthMode::ServerOnly => None,
```

Comment:
This seems weird since we have a case where the ServerOnly, does the client enable Mutual and the Server enable this? And what if neither authenticates? Both are ServerOnly? We should nail down consistent naming here :)

Reply:
why would the client dial someone they don't want to talk to? Noise IK already provides this authentication for the client w/o needing a specific trusted_peers set.